README.md
# [TheStargateProject.com](https://TheStargateProject.com)
Software for Kristian's Fully Functional 3D Printed Stargate

## Setup Instructions:
It is _highly_ recommended to build your gate by using the pre-built Disk Image (ISO) provided by Kristian.
 - If you are **upgrading an existing gate** read this first: [UPGRADING_FROM_V3.X.md](UPGRADING_FROM_V3.X.md)
 - EXPRESS SETUP process can be found in [EXPRESS_SETUP.md](EXPRESS_SETUP.md)
 - A manual setup guide can be found in [ENVIRONMENT_SETUP.md](ENVIRONMENT_SETUP.md)

## Default SSH/SCP credentials
```
Username: pi
Password: sg1
```

## Start/Stop/Restart the Stargate Software manually
The Stargate Software will automatically start when the Raspi boots. It runs as a systemd daemon called `stargate.service`. If you want to manually start/stop/restart it, you can use these commands:
```
sudo systemctl start stargate.service
sudo systemctl stop stargate.service
sudo systemctl restart stargate.service
```

## Web interface
A web interface is provided to allow testing of individual hardware components, dialing, address book, and much more. Find it here:

[http://stargate.local](http://stargate.local)

## Web API
There is a JSON API to interact with the Stargate via a web service. This enables anyone to build software that can control the Stargate. We'd love to see someone build a Stargate Command computer-style controller, or a RPG game that interfaces with the 'Gate. Your imagination is the limit!

The documentation can be found in the repo, or at one of the below links

- v1.0.0 (Current): https://app.swaggerhub.com/apis-docs/TheStargateProject/StargateWebAPI/1.0.0#/
- v1.1.0 (In development): https://app.swaggerhub.com/apis-docs/TheStargateProject/StargateWebAPI/1.1.0

## Credits
- Kristian Tysse designed and wrote all of the original code, most of which is still in use today's program.
- Jonathan Moyes restructured the code and extended it to include additional functionalities.
- The Web UI and basic implementation of the Stargate API Server were based on Dan Clarke's work: https://github.com/danclarke/WorkingStargateMk2Raspi

Stargate SG-1, Stargate Atlantis & Stargate Universe are ™ & © of Metro-Goldwyn-Mayer Studios Inc.  This project is in no way sponsored or endorsed by: SyFy or MGM. This project was created solely as a hobby project and to help other Stargate fans create their own Stargates and to keep the passion and love for Stargate alive.

TheStargateProject.com is a fan-based project and is not intended to infringe upon any copyrights or registered trademarks.

# Development
## Running PyLint
 - To run for StargateMilkyWay: `pylint --rcfile=.pylintrc-milkyway ./*`
 - To run for StargatePegasus:  `pylint --rcfile=.pylintrc-pegasus ./*`


changelog.md
## New in version 4.0.0
This version is a major upgrade from the v3.x branch. Completely refactored and with a lot of exciting new functionality and opportunities to extend the existing functionality!

- Vastly improved setup and installation process - quick and easy!
- Added web interface to dial, see what planet an incoming wormhole is from, run tests, and more
  - Dialing by pressing symbol buttons in a web browser
  - Dialing by clicking on an entry in the Address Book (both Movie Gates, and Fan Gates)
  - Shows what type of wormhole (incoming/outgoing), if any, and what planet is connected
  - Shows the approximate time-to-disengage (~38 minutes max)
  - Dialing a black hole is dangerous - you shouldn't do that. But if you did it's dangerous. You've been warned!!
  - Testing functions:
    - Cycle Chevrons individually
    - Move the symbol ring
    - Turn on ALL of the chevron LEDs
    - Turn off ALL of the LEDs on the Gate
    - Instantly open a wormhole
    - Close any open wormhole
    - Simulate an incoming wormhole (most requested feature)
    - Enter a DHD Test Mode, wherein pressing a button will toggle it's LED on/off
  - Administrative functions:
    - Volume up/down (+/-5%/click) Volume level is saved and persists between reboots
    - Restart the Stargate Software
    - Reboot the Raspberry Pi
    - Shutdown the Raspberry Pi (no more dirty shutdowns!)
  - In-App Subspace Network configuration: no more typing commands in SSH!
  - Shows helpful debug information, internet & subspace connection status, and software update information.
  - Responsive Web UI (Bootstrap+JQuery) UI works well on Mobile Devices, too.
  - Configuration of all aspects of the software is available in the web interface.
    - Type and value validation of all input data ensures system stability
  - A Symbol Overview page details all of the symbols on the gate/DHD, their keyboard mappings and symbol numbers, and provides links to Wikipedia pages for more information about each constellation
  - Helpful FAQs and links to resources for building and troubleshooting your gate
  - A System Information page provides information about the Raspi and Stargate Software running
  - Lifetime Statistics track the number of failed dialing attempts, count of inbound/outbound wormholes established to different types of gates, the amount of time connected.
- The software will run as normal even if hardware isn't connected or is not connected correctly. The gate will gracefully degrade to a "hardware simulation" state, where everything works as it should, just without actually controlling the hardware.
- Startup and shutdown are now graceful, and the software cleans up it's threads on exit.
- Completely restructured the code for readability, adherence to industry/language standards, and facilitating further extension of the software.
- Configurations have been moved out of static Python code into JSON configuration files to allow easy, dynamic reconfiguration all in one place
- Migrated to an HTTP-based API for retrieving Fan Gates and Software Updates - previously done via SQL server direct query.
- Fan Gates are automatically updated every 30 minutes. Previously required a reboot.
- Software Updates are checked every 6 hours. Previously required a reboot.
- Software now runs as a systemd daemon. Manage the daemon with these commands:
```
sudo systemctl start stargate.service
sudo systemctl stop stargate.service
sudo systemctl restart stargate.service
sudo systemctl status stargate.service
```
- Added `avahi` / `Bonjour` so the 'Gate can be accessed as `stargate.local` - you don't need to know it's IP address anymore!
- Adopted SemVer versioning
- Fully documented Web/HTTP API
- Code is linted with pylint
- Lots of behind the scenes stuff to make development, extension and debugging easier
- Installation script and pre-prepared disk image (for the Raspi 3B+) make installation a breeze
- Adds the Stargate Loopback Address (27, 7, 15, 1, 2), analogous with 127.0.0.1 to allow simulating incoming wormholes
- Security enhancements
- Better exception handling to prevent crashes
- Improved memory management
- Added firewall (UFW) to limit access. This is redundant and complementary to the firewalling that occurs on the Subspace Network.
- Improved logging - threads don't step on STDOUT anymore
- Added logrotate, so log files don't get huge and slow things down or fill up the drive. Retains 30 days of daily log file backups, FIFO.
- Automatic configuration of the audio devices to prevent snafus.
- Multi-universe support

## FEATURES on the backlog for possible future development
  - Scheduled Iris open/closed times
  - Scheduled mute times
  - Button in Web to open/close the Iris
  - Button in Web to connect/disconnect from subspace
  - Incoming wormhole alert in Web
  - Configure WiFi via Bluetooth (WebBluetooth in a browser), so you can skip the wpa_supplicant part of setup.
  - Implement homing
  - Support for the Milky Way Main Board
  - More Web stuff: fun interfaces, alternate "keyboard layouts," hidden movie references, etc.
  - Upload your own, additional sound clips
  - Play sound clips by clicking buttons in the Web UI
  - "Call Log" in Web: Shows last 25 incoming and outgoing wormholes, duration established, and what planet.
  - Show log files in Web (WebSocket streaming?)
  - Consider moving API away from polling to use WebSockets? Polling was just the quick/easy way out.
  - More granularity in hardware fall-back modes

## New in version 3.7:
- The Stargate server will now also start with eth0 interface only. (If subspace or wlan is missing)

## New in version 3.6:
- DHD detection update. Planet/stargate name is used in the log instead of the IP.

## New in version 3.5:
- Fix an issue with audio output. The program will now edit the alsa.conf file to set the USB audio adapter as output.


classes/DHD.py
class DHD:
    """
    This is the old DHD version. Not recommended. Use the new version instead.
    """
    def __init__(self):
        import adafruit_dotstar as dotstar
        import board
        self.dots = dotstar.DotStar(board.D14, board.D15, 39, brightness=0.01)
        self.center_dot = dotstar.DotStar(board.D14, board.D15, 1, brightness=0.15)

    def light_on(self, symbol_number, color):
        """
        This helper function activates the light for the dhd button of "symbol" with the "color" specified.
        :param symbol_number: The symbol number of which to control.
        :param color: The color for the led as a tuple. eg: (250, 117, 0)
        :return: nothing is returned.
        """
        symbol_number_to_dhd_light_map = {0: 0, 1: 25, 2: 19, 3: 38, 4: 20, 5: 23, 6: 30, 7: 28, 8: 3, 9: 22,
                                          10: 11, 11: 36, 12: 34, 14: 17, 15: 6, 16: 9, 17: 18, 18: 16, 19: 26,
                                          20: 21, 21: 37, 22: 27, 23: 15, 24: 29, 25: 1, 26: 14, 27: 4, 28: 10, 29: 31,
                                          30: 8, 31: 5, 32: 24, 33: 12, 34: 33, 35: 7, 36: 2, 37: 35, 38: 32, 39: 13}
        self.dots[symbol_number_to_dhd_light_map[symbol_number]] = color
    def lights_off(self):
        """
        This method turns off all dhd lights
        """
        self.dots.fill((0, 0, 0))
    def center_light_on(self):
        """
        A helper function to turn on the centre dhd light. It needs a bit more brightness than the other lights. Hence
        this extra method.
        """
        self.center_dot[0] = (255, 0, 0)

import PyCmdMessenger
from pprint import pprint

class DHDv2:

    def __init__(self, port, baud_rate):
        # Initialize an ArduinoBoard instance.
        self.board = PyCmdMessenger.ArduinoBoard(port, baud_rate=baud_rate)

        # List of command names (and formats for their associated arguments). These must
        # be in the same order as in the sketch.
        self.commands = [
            ["get_fw_version", "s"],
            ["get_hw_version", "s"],
            ["get_identifier", "s"],
            ["reset", ""],
            ["evt_error", "s"],
            ["evt_ack", ""],

            ["message_bool", "?"],
            ["message_string", "s"],
            ["message_int", "i"],
            ["message_long", "l"],
            ["message_double", "d"],
            ["message_color", "iii"],

            ["clear_all", ""],  # Turns off all pixels in the RAM buffer.
            ["clear_pixel", "i"],
            # Turns off the pixel at the provided index, in the RAM buffer. [ pixelIndex ]. Index starts at 0.
            ["set_all", "iii"],  # Sets all pixels to color provided, in the RAM buffer. [ red, green, blue ]
            ["set_pixel", "iiii"],
            # Sets pixel to color provided, in the RAM buffer.[ pixelIndex, red, green, blue ]. Index starts at 0.

            ["get_pixel_count", "i"],  # Returns the number of pixels managed by the library instance

            ["set_brightness_symbols", "i"],  # Set brightness for the symbol rings' LEDs.
            # ** Per library documentation: "Intended to be called once, in setup(),
            # to limit the current/brightness of the LEDs throughout the life of the
            # sketch. It is not intended as an animation effect itself! The operation
            # of this function is “lossy” — it modifies the current pixel data in RAM,
            # not in the show() call — in order to meet NeoPixels’ strict timing
            # requirements. Certain animation effects are better served by leaving the
            # brightness setting at the default maximum, modulating pixel brightness
            # in your own sketch logic and redrawing the full strip with setPixel()."
            ["set_brightness_center", "i"],  # Set brightness for the center LED. Call this once: cautions as above.

            ["latch", ""],  # Transmits the current pixel configuration in the RAM buffer out to the pixels
        ]

        # Initialize the messenger
        self.c = PyCmdMessenger.CmdMessenger(self.board, self.commands)

        pass

    def getFirmwareVersion(self):
        self.c.send("get_fw_version")
        return self.c.receive()[1][0]

    def getHardwareVersion(self):
        self.c.send("get_hw_version")
        return self.c.receive()[1][0]

    def getIdentifierString(self):
        self.c.send("get_identifier")
        return self.c.receive()[1][0]

    def getPixelColorTuple(self, pixelIndex):
        self.c.send("get_pixel_color", pixelIndex)
        pprint(self.c.receive())
        return self.c.receive()[1]

    def getPixelCount(self):
        self.c.send("get_pixel_count")
        return self.c.receive()[1][0]

    def setBrightnessSymbols(self, brightness):
        self.c.send("set_brightness_symbols", brightness)
        return True

    def setBrightnessCenter(self, brightness):
        self.c.send("set_brightness_center", brightness)
        return True

    def setAllPixelsToColor(self, red, green, blue):
        self.c.send("set_all", red, green, blue)
        return True

    def setPixel(self, pixelIndex, red, green, blue):
        symbol_number_to_dhd_light_map = {0: 0, 1: 34, 2: 2, 3: 21, 4: 20, 5: 36, 6: 29, 7: 31, 8: 18, 9: 37,
                                          10: 10, 11: 23, 12: 25, 14: 4, 15: 15, 16: 12, 17: 3, 18: 5, 19: 33,
                                          20: 38, 21: 22, 22: 32, 23: 6, 24: 30, 25: 1, 26: 7, 27: 17, 28: 11, 29: 28,
                                          30: 13, 31: 16, 32: 35, 33: 9, 34: 26, 35: 14, 36: 19, 37: 24, 38: 27, 39: 8}
        self.c.send("set_pixel", symbol_number_to_dhd_light_map[pixelIndex], red, green, blue)
        return True

    def setPixel_use_LED_id(self, pixelIndex, red, green, blue):
        self.c.send("set_pixel", pixelIndex, red, green, blue)
        return True

    def clearAllPixels(self):
        self.c.send("clear_all")
        return True

    def clearPixel(self, pixelIndex):
        self.c.send("clear_pixel", pixelIndex)
        return True

    def latch(self):
        self.c.send("latch")
        return True

class KeyboardMode:
    """
    This is just a fallback class that disables all the DHD LED functions in case the DHD is not present. You can use a regular keyboard instead.
    To dial Aphopis's base, just hit cFX1K98A on your keyboard.
    """

    def __init__(self):
        pass

    def getFirmwareVersion(self):
        pass

    def getHardwareVersion(self):
        pass

    def getIdentifierString(self):
        pass

    def getPixelColorTuple(self, pixelIndex):
        pass

    def getPixelCount(self):
        pass

    def setBrightnessSymbols(self, brightness):
        pass

    def setBrightnessCenter(self, brightness):
        pass

    def setAllPixelsToColor(self, red, green, blue):
        pass

    def setPixel(self, pixelIndex, red, green, blue):
        symbol_number_to_dhd_light_map = {0: 0, 1: 34, 2: 2, 3: 21, 4: 20, 5: 36, 6: 29, 7: 31, 8: 18, 9: 37,
                                          10: 10, 11: 23, 12: 25, 14: 4, 15: 15, 16: 12, 17: 3, 18: 5, 19: 33,
                                          20: 38, 21: 22, 22: 32, 23: 6, 24: 30, 25: 1, 26: 7, 27: 17, 28: 11, 29: 28,
                                          30: 13, 31: 16, 32: 35, 33: 9, 34: 26, 35: 14, 36: 19, 37: 24, 38: 27, 39: 8}
        pass

    def setPixel_use_LED_id(self, pixelIndex, red, green, blue):
        pass

    def clearAllPixels(self):
        pass

    def clearPixel(self, pixelIndex):
        pass

    def latch(self):
        pass
